### PR TITLE
catalog builds: table ordering refactor, and improved implicit projections

### LINKED ISSUE #231: inference: missing implicit projections for collection key within `additionalItems` array
When using a schema that composes `additionalItems` and `minItems`, where the collection key references an index < minItems, we fail to generate a projection for the collection's key. This fails later, during validation of the CollectionSpec.

For example given collection schema:
```yaml
type: object
properties:
  arr:
    type: array
    additionalItems: {type: integer}
    minItems: 2
required: [arr]
```
And collection key `[/arr/1]`, we (correctly) allow the collection key because we infer its required existence and type.
But we don't create an implicit projection having this location. Currently, downstream bits (and the Go `flow.CollectionSpec` implementation) require that it do exist (and it seems pretty reasonable that it would).

### PR DESCRIPTION
**Pull request by jgraettinger**
Definitely review this commit by commit. They tell the story of how this change set came together.

It was a bit of a rabbit hole to fix a seemingly minor bug. I found myself really, really wanting reliable guarantees that tables were ordered by their natural keys, that could be leveraged for lookups throughout the validation codebase. It's a little more complicated than simply slapping a BTreeSet into the mix, because our tables are actually _multi_sets. We want to represent multiple instances of various entities, so that we can generate good catalog build errors as we walk those duplicates.

Fixes #231

---
This change is [Reviewable](https://reviewable.io/reviews/estuary/flow/232)

### CODE REVIEW

File: crates/models/src/tables/macros.rs [hunk 1]
```diff
@@ -318,13 +348,15 @@ macro_rules! tables {
                 let mut stmt = db.prepare(&Self::select_sql())?;
                 self.0.extend(stmt.query_map([], $row::scan)?
                               .collect::<Result<Vec<_>, _>>()?);
+                self.reindex();
```
Comment (saterus) [inline crates/models/src/tables/macros.rs:351]: I think you can replace `self.0.extend` + `self.reindex` with `self.extend` now, right?

File: crates/models/src/tables/macros.rs [hunk 2]
```diff
@@ -318,13 +348,15 @@ macro_rules! tables {
...
                 Ok(())
             }
 
             fn load_where(&mut self, db: &rusqlite::Connection, filter: &str, params: &[&dyn rusqlite::types::ToSql]) -> rusqlite::Result<()> {
                 let mut stmt = db.prepare(&format!("{} WHERE {}", Self::select_sql(), filter))?;
                 self.0.extend(stmt.query_map(params, $row::scan)?
                               .collect::<Result<Vec<_>, _>>()?);
+                self.reindex();
```
Comment (saterus) [inline crates/models/src/tables/macros.rs:359]: Ditto.

**Review by saterus — APPROVED** [review]
After spending a lot of time reading through the catalog build process, I understand the problem and why this fixes it. :+1: 

The refactorings seem like a win too. After looking into `superslice`, I understand why you were keen to have the tables maintain a sorted ordering.

File: crates/models/src/tables/macros.rs (hunk 1 above)
Comment (jgraettinger) [reply crates/models/src/tables/macros.rs]: Done.

File: crates/models/src/tables/macros.rs (hunk 2 above)
Comment (jgraettinger) [reply crates/models/src/tables/macros.rs]: Done

**Pull request merged**